### ISSUE TITLE
Past spikes from Google Analytics history and how to handle them

### ISSUE THREAD

**Issue by wholesomedev** [hollowverse-archive/hollowverse#151]
I looked at the past traffic patterns of Hollowverse and noticed the following:

### [Aug 12 2014](https://analytics.google.com/analytics/web/#report/visitors-overview/a5715214w55386753p56394045/%3F_u.date00%3D20140812%26_u.date01%3D20140812%26overview-graphOptions.selected%3Danalytics.nthHour/)

In one hour traffic went from 500 to over 7,000.

### [Apr 21 2016](https://analytics.google.com/analytics/web/#report/visitors-overview/a5715214w55386753p56394045/%3F_u.date00%3D20160420%26_u.date01%3D20160423%26overview-graphOptions.selected%3Danalytics.nthHour/)

In about 2 hours, traffic went from 900 to over 8,000.

### [Feb 6 2017](https://analytics.google.com/analytics/web/#report/visitors-overview/a5715214w55386753p56394045/%3F_u.date00%3D20170205%26_u.date01%3D20170207%26overview-graphOptions.selected%3Danalytics.nthHour/)

Within an hour, traffic went from 3000 to 10,000.

----

## Flexible vs Standard

I read this page which has the description of the difference between the [flexible and the standard environment](https://cloud.google.com/appengine/docs/the-appengine-environments). It doesn't seem to me that it is possible for us to use the standard environment at all because it only supports specific programming languages, Node.js is not one of them. And obviously, we don't want to use pure Compute Engine instances because that would mean we have to manually manage our infrastructure.

It seems to me that we're stuck with the flexible environment. I hope Google will improve the experience of the flexible environment so that it is able to scale more quickly.

## Increasing instance resources

In light of our recent discovery of how slow new instances get deployed, it seems to me that increasing the CPU and RAM of the instances may alleviate some of the impact of sudden traffic spikes because each newly deployed instance will be able to hand that much more traffic. I'm not sure how accurate this assumption is though.

We'll probably use a CDN for the static assets, but we cannot use a CDN for the entire page as is being done with the current Hollowverse because the new Hollowverse is much more dynamic.

Anyway, I just wanted to share these thoughts. Not much action we can take at this point other than increasing the CPU and RAM of the instances. Let me know what you think.

**Comment by forabi** [comment]
> In light of our recent discovery of how slow new instances get deployed, it seems to me that increasing the CPU and RAM of the instances may alleviate some of the impact of sudden traffic spikes because each newly deployed instance will be able to hand that much more traffic. I'm not sure how accurate this assumption is though.

We can increase the number of cores used in an instance, but not the performance of a single core. Like I stated in a earlier discussion on Discord, Node.js is mostly single-threaded, so adding more cores does not yield any significant improvements in performance. Even single-core performance is not usually the bottleneck, because network and I/O are usually slower. To increase the number of requests a single Node.js process can handle, we need to avoid blocking the event loop as much as possible, which is already taken care of by delegating processing of network and I/O events to native modules and using callbacks to inform JS of the results.

So the only way to scale Node.js is to use multiple processes, and this is why it is usually used behind a reverse proxy, like NGINX, which spins up a number of Node.js processes and balances the load between these processes dynamically. This is similar to what App Engine does, and in fact, App Engine uses NGINX internally.

I suggest we use a 3rd-party service to perform load tests on the currently deployed instances and simulate the traffic patterns on the dates you mentioned. We can then observe how App Engine scales and if we need to do something to improve the performance.

**Comment by wholesomedev** [comment]
> I suggest we use a 3rd-party service to perform load tests on the currently deployed instances and simulate the traffic patterns on the dates you mentioned. We can then observe how App Engine scales and if we need to do something to improve the performance.

Yeah, that's a good idea. We should do it. I'll do some more research on this.

**Comment by wholesomedev** [comment]
@forabi I did some research. I looked into loadimpact.com and loader.io. 

loadimpact.com is a bit over our price range. loader.io seems adequate. Can you try to perform some load tests using loader.io to mimic the traffic patterns linked above?

**Comment by forabi** [comment]
@msafih I've looked at loadimpact.com and loader.io. The latter offers a very limited test duration for trial accounts. I found https://flood.io/ and it seems to be good for our use case. It offers 5 hours for trial accounts and has a nice UI.

I will start a simple test with 200 users for 5 minutes and increase the load gradually. After a cool down period, I will perform a larger tests with thousands of virtual users for 1 hour and see how App Engine scales.
